### PR TITLE
Additional WebDAV features: status, get properties, set properties, filter

### PR DESCRIPTION
**Pull request by fkleon**
Hi,

I've implemented some additional WebDAV features that I needed for my project.

New features:
* Expose additional fields on `WebDavFile`, including the unique file ID
* `status` method to query WebDAV server capabilities; can also be used to check the connection to the server
* `getProps` method to retrieve selected (or all) properties for a resource
* `updateProps` method to set properties on a resource
* `filter` method to run the [filter-files](https://docs.nextcloud.com/server/19/developer_manual/client_apis/WebDAV/basic.html#listing-favorites) report, e.g. to list all favourites of a user
* Support for properties with custom namespaces on `ls`, `getProps` and `updateProps`

Improvements:
* Ability to specify which properties should be requested during `ls`
* Improve directory detection to work with collection resources
* Add optional `overwrite` parameter to `copy`and `move`

I don't think there are any backwards incompatible (breaking) changes in here.

The commits are pretty well organised but I can also split this up into separate pull requests if you prefer smaller changesets.

### CODE REVIEW

**Comment by provokateurin** [comment]
Sorry I currently have little time to review PRs. Please keep submitting them. I hope I can check them in the next days

**Comment by provokateurin** [comment]
Thanks for this huge contribution. I looked through everything and everything looks very well.
One thing I was thinking about was that it may be a good idea to store the prop names in variables so that for example I can access `nc:sharees` using `WebDavClient.props.ncSharees`. Then `WebDavClient.allProps` would be acces through `WebDavClient.props.all`. What do you think?

**Comment by fkleon** [comment]
@jld3103 Thanks for reviewing this so quickly.

Yes, I agree, exposing the available properties as constants make a lot of sense - I found myself misspelling properties on multiple occasions. I'll rework this part, probably will get to it in the next few days.

**Comment by fkleon** [comment]
@jld3103 I've updated this PR, have a look.

File: lib/src/webdav/file.dart [hunk 1]
```diff
@@ -1,48 +1,184 @@
...
 
   // ignore: public_member_api_docs
-  final List<int> shareTypes;
+  List<int> shareTypes = [];
+
+  /// User IDs of sharees.
+  List<String> sharees = [];
+
+  /// Whether this file is marked as favorite.
+  bool favorite = false;
```
Comment (provokateurin) [inline lib/src/webdav/file.dart:53]: Is there a reason why these fields can't be final?

File: test/nextcloud_test.dart [hunk 2]
```diff
@@ -114,27 +122,173 @@ void main() {
...
+          'true');
+    });
+    test('Filter files', () async {
+      final path = '${config.testDir}/filter-test.txt';
+      final data = utf8.encode('WebDAV filtertest');
+      final response = await client.webDav.upload(data, path);
+      final id = response.headers['oc-fileid'];
+
+      // Favorite file
+      await client.webDav.updateProps(path, {'oc:favorite': '1'});
```
Comment (provokateurin) [inline test/nextcloud_test.dart]: This should use the constant (there are more occurences below)

**Comment by provokateurin** [comment]
Small things, but otherwise LGTM

File: lib/src/webdav/file.dart (hunk 1 above)
Comment (fkleon) [reply lib/src/webdav/file.dart:53]: I'm building the `WebDavFile` iteratively from XML and didn't want to introduce an intermediate builder object.
Given that many fields are optional it also allows for a more intuitive way of defining defaults.

Is there a particular reason you're wanting to keep all of this final?

File: test/nextcloud_test.dart (hunk 2 above)
Comment (fkleon) [reply test/nextcloud_test.dart]: Good point, updated. I left a few non-constant uses in place in the tests to emphasise that it's just a string.

**Comment by provokateurin** [comment]
Thanks for the contribution

**Comment by provokateurin** [comment]
Published as `1.3.0`

**Pull request merged**